spyder/plugins/explorer/tests/test_explorer.py

```diff
@@ -46,5 +66,99 @@ def test_project_explorer(project_explorer):
     assert project_explorer
 
 
+project_dir = osp.join(
+        getcwd_or_home(),
+        'temp_dir_test_file_explorer_functions_copy_save_path_file')
+if not osp.exists(project_dir):
+        os.mkdir(project_dir)
+project_file1 = osp.join(project_dir, 'script.py')
+with open(project_file1, 'w') as fh:
+    fh.write('Spyder4 will be released this year')
+project_file2 = osp.join(project_dir, 'pyscript.py')
+with open(project_file2, 'w') as fh:
+    fh.write('Spyder4')
+subdir = osp.join(project_dir, 'subdir')
+if not osp.exists(subdir):
+    os.mkdir(subdir)
+project_file3 = osp.join(subdir, 'Columbia.txt')
+with open(project_file3, 'w') as fh:
+    fh.write('South America')
+cb = QApplication.clipboard()
```

Comment:
Everything you added as part of the "test" function here really should be part of what is currently called the ``copy_path_files`` fixture, since it only serves to be referenced from other tests and has nothing to do with this test itself.

Instead, a better solution is to parameterize the *fixture* over the file paths instead of the below test functions, with it generating a project explorer opened to a temp directory with the files at each of a list of paths from the parameterized variable (including any sub-directories) created with procedurally generated contents (e.g. "This is a file at path ``file_path``"). Then, Pytest will automatically run each of your tests over each instance of the fixture. You also want to return the paramartized variable containing the paths for that particular case from your ``copy_path_files`` fixture, so your tests know what to expect.

Reply:
To revise the above, ideally the creation of the temp directory and the files/directories at the specified paths with boilerplate content and its accompanying paramaterization and teardown code (cleaning the directory) should be a separate fixture from the ``copy_path_files`` one, since its distinct from that, and have *that* fixture return the list of file paths.

Also, if you need different files/directories to be tested per test function, then you'll need to instead paramaterize the paths over each function (as you have now, just with the paths themselves rather than referring to already created file objects) and then use a [fixture factory](https://docs.pytest.org/en/latest/fixture.html#factories-as-fixtures) to create them, but that adds complexity so if you can get away with testing the same sets of files/dirs for each test function, that would be ideal.

Reply:
I included them into fixture function and removed mark.parametrize.

Reply:
Okay, good. Two comments:

First, as I mentioned, you should *parameterize* your fixture over the paths, i.e. include a ``pytest.mark.parameterize`` before the fixture with the paths as lists of strings like you did for the test functions,  rather than hardcoding all of them in the fixture itself.

Second, I requested you make the file contents procedurally generated (e.g. text like ``"This is the file with path: {file_path}\n".format(file_path=str(path)) rather than bespoke and thus requiring tons of branching, hard-coded logic to handle.

Together, these two changes should greatly simplify and improve both your fixture and your test functions and eliminate the need for most of the remaining branching and hard-coding in both. Your fixture would just consist of a for loop that, for each ``path`` in the list from ``pytest.mark.paramaterize``, creates any intervening directories with ``os.makedirs(exist_ok=True)`` and either a file with the above procedurally-generated contents or, if the path ends in ``/``, a directory. Then, your test function code just iterates over those paths, with expected results for ``isfile()`` and ``isdir()``, the path, the number of paths and the file contents if a file determined and checked as appropriate.

Reply:
I applied your suggestions. Please check if it is resolved!